[Essentials] Geolocation foreground listening

vividos opened the pull request:
### Description of Change

This PR implements Geolocation foreground listening, Spec was suggested in #3678, and original Xamarin.Essentials issues are https://github.com/xamarin/Essentials/issues/1447 and https://github.com/xamarin/Essentials/issues/290

As I never got feedback from the Spec issue tickets, I'm creating a PR with my changes sitting on my HD for too long. I'm happy to discuss and change any part of the code, refactor, etc.

I'm targeting `main` branch, which I guess goes into .NET 7, which is fine for me. Just tell me when I should switch base.

### Issues Fixed

Fixes #3678

vividos committed:
added geolocation foreground listening public API without implementation

vividos committed:
added section on geolocation sample page for foreground listening feature

vividos committed:
implemented geolocation foreground listening for Android, iOS, MacOS and UWP

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/maui?pullRequest=9572) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/maui?pullRequest=9572) before we can accept your contribution.<br/><br/>:x: vividos [sign now](https://cla.dotnetfoundation.org/dotnet/maui?pullRequest=9572)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/maui?pullRequest=9572) it.</sub>

rmarinho commented:
/azp run

vividos commented:
Hmm how do I get rid of the `RS0016` errors? Visual Studio's quick action menu added the signatures to the `PublicAPI.Unshipped.txt` file, but it doesn't help...

inforithmics commented:
Fantastic, this is the last missing functionality of https://github.com/jamesmontemagno/GeolocatorPlugin that is not in Maui.Essentials.

vividos committed:
moved common code for determining DesiredAccuracy on UWP to extension method

vividos commented:
I did a minor refactoring of duplicated code, but otherwise this PR is ready to review.

vividos commented:
@mattleibow @rmarinho @Eilon What are the next steps for this PR? Would you like to discuss the public API, or review the code? What about documentation? Ask someone for the Tizen implementation?

vividos commented:
@jfversluis @davidortinau I have an open feature here that basically waits to be discussed, reviewed and finally committed since about two years, first in the Xamarin Essentials repository, and now here for MAUI essentials. Any word on how to proceed with the feature would be great from you, even a "no we don't do that".
The feature is (for two of my projects, one commerical), and surely for other people, the last thing that remains from James Montemagno's plugin projects to be moved to Essentials. Every other plugin has now a more modern counterpart, being Essentials, Shiny or other well maintained community projects.
So please say anything, thanks!

SokoFromNZ commented:
Any news here? I would really need this feature in MAUI!

vividos commented:
Still waiting for any comment from MAUI folks, see my comment from Oct 1....

TFreudi1 commented:
I'm also interested in this  listner !

orwo1 commented:
Any progress on this?
So long that this wasn't even inserted into Xamarin Essentials, and now it won't even be inserted into Maui?

jfversluis commented:
Hey @vividos somehow this one fell through the cracks at least for me, sorry about that. It has my attention now!

Are you still up for bringing this home? Then I will make sure this gets the attention it deserves.

jfversluis commented:
/azp run

vividos commented:
@jfversluis definitely! First step would be to check the API, which is provided in #3678. If that is ok, there would be some changes necessary on this PR, as interfaces were introduced some time ago, and I see that XML comments can be in code now. I can work on that from 22nd Jan on.
More important is if the API is in tune with what MAUI Essentials wants to use.
Thanks Gerald for picking this up!

jfversluis commented:
@vividos added some notes about the API design in that linked issue. Nothing too shocking. Let's get those details out of the way and then let's add that to here and get things going.

vividos commented:
Thanks Gerald, I made a comment on the Spec issue #3678. I can work on the Spec and code from Jan 22nd on. If anyone wants to jump in and continue working on the PR, feel free!

vividos committed:
renamed class ListeningRequest to GeolocationListeningRequest

vividos committed:
renamed property IsListening to IsListeningForeground

vividos committed:
added xml documentation for all new public methods and properties

vividos committed:
implemented event LocationError, using enum GeolocationError and class GeolocationErrorEventArgs

vividos committed:
fixed potential leak where ContinuousLocationListener keeps the reference to the GeolocationImplementation on iOS

vividos committed:
changed StopListeningForegroundAsync() to StopListeningForeground() and return void

vividos committed:
fixed error in Essentials samples where async keyword is not necessary anymore

vividos committed:
enabled nullable checks for GeolocationListeningRequest class

vividos commented:
@jfversluis I think this PR is ready for review now.

jfversluis commented:
/azp run

vividos committed:
renamed ListeningRequest.ios.macos.cs to match class name; no source code changes

jfversluis commented on src/Essentials/src/Geolocation/Geolocation.android.cs:
```diff
@@ -112,6 +117,84 @@ void RemoveUpdates()
 
+		public async Task<bool> StartListeningForegroundAsync(GeolocationListeningRequest request)
+		{
+			_ = request ?? throw new ArgumentNullException(nameof(request));
```
```suggestion
			ArgumentNullException.ThrowIfNull(nameof(request));
```

vividos replied:
I checked the Maui project, and `ArgumentNullException.ThrowIfNull()` is not used much, but I will do the change in Geolocation. I like it, it's more elegant.

vividos replied:
It seems the syntax actually is `ArgumentNullException.ThrowIfNull(request);` and the nameof(request) in the second parameter can be omitted and is derived from the first parameter.

jfversluis commented on src/Essentials/src/Geolocation/Geolocation.android.cs:
```diff
@@ -112,6 +117,84 @@ void RemoveUpdates()
 			}
 		}
 
+		public async Task<bool> StartListeningForegroundAsync(GeolocationListeningRequest request)
```
Could you add code documentation for this? Basically just add three slashes (`///`) at the top and start filling out what Visual Studio proposes, also see: https://github.com/dotnet/maui/blob/main/docs/CodeDocumentationGuidelines.md and here for some examples: https://github.com/dotnet/maui/pulls?q=is%3Apr+author%3Ajfversluis+label%3A%22area%2Fdocs+%F0%9F%93%93%22+is%3Aclosed

jfversluis commented on src/Essentials/src/Geolocation/Geolocation.android.cs:
```diff
@@ -112,6 +117,84 @@ void RemoveUpdates()
+			}
+		}
+
+		public void StopListeningForeground()
```
Also here, please add code comment documentation

jfversluis commented on src/Essentials/src/Geolocation/Geolocation.ios.macos.cs:
```diff
@@ -91,6 +96,77 @@ void Cancel()
 			}
 		}
+
+		public async Task<bool> StartListeningForegroundAsync(GeolocationListeningRequest request)
```
Code documentation, see earlier comments

jfversluis commented on src/Essentials/src/Geolocation/Geolocation.ios.macos.cs:
```diff
@@ -91,6 +96,77 @@ void Cancel()
+
+		public async Task<bool> StartListeningForegroundAsync(GeolocationListeningRequest request)
+		{
+			_ = request ?? throw new ArgumentNullException(nameof(request));
```
```suggestion
			ArgumentNullException.ThrowIfNull(nameof(request));
```

jfversluis commented on src/Essentials/src/Geolocation/Geolocation.ios.macos.cs:
```diff
@@ -91,6 +96,77 @@ void Cancel()
+			}
+		}
+
+		public void StopListeningForeground()
```
Code documentation, see earlier comments

jfversluis commented on src/Essentials/src/Geolocation/Geolocation.ios.macos.cs:
```diff
@@ -116,4 +192,36 @@ public override void LocationsUpdated(CLLocationManager manager, CLLocation[] lo
+
+		internal Action<GeolocationError> ErrorHandler { get; set; }
+
+		public override void LocationsUpdated(CLLocationManager manager, CLLocation[] locations)
```
Code documentation, see earlier comments.

Since this is an override you can probably just do `/// <inheritdoc/>`

jfversluis commented on src/Essentials/src/Geolocation/Geolocation.ios.macos.cs:
```diff
@@ -116,4 +192,36 @@ public override void LocationsUpdated(CLLocationManager manager, CLLocation[] lo
+			LocationHandler?.Invoke(location);
+		}
+
+		public override void Failed(CLLocationManager manager, NSError error)
```
Code documentation, see earlier comments.

Since this is an override you can probably just do `/// <inheritdoc/>`

jfversluis commented on src/Essentials/src/Geolocation/Geolocation.ios.macos.cs:
```diff
@@ -116,4 +192,36 @@ public override void LocationsUpdated(CLLocationManager manager, CLLocation[] lo
+				ErrorHandler?.Invoke(GeolocationError.PositionUnavailable);
+		}
+
+		public override void AuthorizationChanged(CLLocationManager manager, CLAuthorizationStatus status)
```
Code documentation, see earlier comments.

Since this is an override you can probably just do `/// <inheritdoc/>`

jfversluis commented on src/Essentials/src/Geolocation/Geolocation.ios.macos.cs:
```diff
@@ -116,4 +192,36 @@ public override void LocationsUpdated(CLLocationManager manager, CLLocation[] lo
+				ErrorHandler?.Invoke(GeolocationError.Unauthorized);
+		}
+
+		public override bool ShouldDisplayHeadingCalibration(CLLocationManager manager) => false;
```
Code documentation, see earlier comments.

Since this is an override you can probably just do `/// <inheritdoc/>`

jfversluis commented on src/Essentials/src/Geolocation/Geolocation.shared.cs:
```diff
@@ -27,6 +29,42 @@ public interface IGeolocation
 		/// <returns>A <see cref="Location"/> object containing current location information or <see langword="null"/> if no location could be determined.</returns>
 		Task<Location?> GetLocationAsync(GeolocationRequest request, CancellationToken cancelToken);
+
+		/// <summary>
+		/// Indicates if currently listening to location updates while the app is in foreground.
+		/// </summary>
```
Yesss perfect! Thank you!

jfversluis commented on src/Essentials/src/Geolocation/Geolocation.tizen.cs:
```diff
@@ -67,5 +69,11 @@ public async Task<Location> GetLocationAsync(GeolocationRequest request, Cancell
 			return lastKnownLocation;
 		}
+
+		public Task<bool> StartListeningForegroundAsync(GeolocationListeningRequest request) =>
+			throw ExceptionUtils.NotSupportedOrImplementedException;
+
+		public void StopListeningForeground() =>
+			throw ExceptionUtils.NotSupportedOrImplementedException;
```
@JoonghyunCho can you or someone on your team look at this maybe?

JoonghyunCho replied:
Oops, I almost missed the comment. Thanks for letting me know!
Currently, there is no actual device target I can run maui and use the real time location service.
Let me leave it as is for now, and see if I can implement with any emulator.
Thanks!

jfversluis commented on src/Essentials/src/Geolocation/Geolocation.uwp.cs:
```diff
@@ -43,16 +47,82 @@ partial class GeolocationImplementation : IGeolocation
+
+		public async Task<bool> StartListeningForegroundAsync(GeolocationListeningRequest request)
+		{
+			_ = request ?? throw new ArgumentNullException(nameof(request));
```
```suggestion
			ArgumentNullException.ThrowIfNull(nameof(request));
```

jfversluis commented on src/Essentials/src/Geolocation/Geolocation.uwp.cs:
```diff
@@ -43,16 +47,82 @@ partial class GeolocationImplementation : IGeolocation
+			}
+		}
+
+		public async Task<bool> StartListeningForegroundAsync(GeolocationListeningRequest request)
```
Code documentation, see earlier comments.

jfversluis commented on src/Essentials/src/Geolocation/Geolocation.uwp.cs:
```diff
@@ -43,16 +47,82 @@ partial class GeolocationImplementation : IGeolocation
+		}
 
-			static void CheckStatus(PositionStatus status)
+		public void StopListeningForeground()
```
Code documentation, see earlier comments.

jfversluis commented on src/Essentials/src/Geolocation/LocationEventArgs.shared.cs:
```diff
@@ -0,0 +1,18 @@
+
+		public LocationEventArgs(Location location)
+		{
+			if (location == null)
+				throw new ArgumentNullException(nameof(location));
```
```suggestion
			ArgumentNullException.ThrowIfNull(nameof(location));
```

vividos replied:
`ArgumentNullException.ThrowIfNull()` is not available with `netstandard2.0` and `netstandard2.1`, so I can't replace the call. I wonder if `location` can ever be `null` with the `#nullable enable` at the start of the file.

jfversluis reviewed (changes requested):
Pretty good! Couple of minor things, mostly add some more code documentation on some public members to make sure we maintain coverage here.

Let me know if you want to add these last bits and I think we're ready to merge, else I will take care of it :)

Thank you so much for your time, effort and mostly patience on this one!

vividos commented:
Thanks Gerald, I'll do the changes, no problem. I love those code docs and always wondered why Essentials didn't have them or moved them out in some XML files.
Another thing that is currently different from James' plugin is the error handler behavior: When UWP signals a `LocationError`, it also calls `StopListeningForeground`, but the other platforms do not. I honestly don't know why I omitted this and will add it, too.

jfversluis commented:
> always wondered why Essentials didn't have them or moved them out in some XML files.

Loooong story there. Same for Forms and now .NET MAUI. We're working on fixing that and actually whole of Essentials should be better by now and have most of it and I saw you added a lot as well so thank you for that. If you're looking for more (easier) contributions in the future, adding/improving that docs stuff is certainly useful.

Anyway, yeah, work through this feedback, add what you found was missing and we'll go from there. Thanks!

vividos committed:
call StopListeningForeground() on Android, iOS and macOS before signalling LocationError event, to make behavior consistent with Windows

vividos committed:
replaced throwing ArgumentNullException with ArgumentNullException.ThrowIfNull()

vividos committed:
added xml documentation for all newly added public geolocaion foreground listening APIs

vividos committed:
removed duplicated code for determining GeolocationAccuracy on iOS and macOS

vividos commented:
@jfversluis I think I addressed all your review comments, except the Tizen implementation. Just add more if you want me to do some more changes. I also took the opportunity to reduce code duplication, see the last commit.
I also have some `#nullable enable` changes for Geolocation on my HD. Should these go into a follow-up PR?
And one more question: Is the `main` branch the correct one? And do you want me to force-push the changes to the latest commit?
Thanks for reviewing this!

jfversluis commented:
/azp run

jfversluis commented:
> I also have some #nullable enable changes for Geolocation on my HD. Should these go into a follow-up PR?

Let's do that, let's focus on getting this in first and then we can refine with what you have for that just so we don't delay on this any further.

> And one more question: Is the main branch the correct one? And do you want me to force-push the changes to the latest commit?

All things go into main and then we look at what we want to backport, so main definitely is fine. We don't have any conflicts it seems and we're squashing commits anyway so I don't think a rebase would add much value, but thank you!

mattleibow commented on src/Essentials/src/Geolocation/LocationEventArgs.shared.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace Microsoft.Maui.Devices.Sensors
+{
+	public class LocationEventArgs : EventArgs
```
This probably should match the event name, so maybe `LocationChangedEventArgs`

vividos replied:
I would rename the class when we decided on the event name and would name it `GeolocationLocationChangedEventArgs` or `GeolocationLocationupdatedEventArgs`, so that all .cs files start with `Geolocation`.

mattleibow replied:
I think `GeolocationLocationChangedEventArgs` is good.

jfversluis replied:
Done!

mattleibow commented on src/Essentials/src/Geolocation/Geolocation.shared.cs:
```diff
@@ -27,6 +29,40 @@ public interface IGeolocation
+		/// Occurs when an error during listening for location updates arises. When getting the event,
+		/// listening for further updates may have been stopped.
+		/// </summary>
+		event EventHandler<GeolocationErrorEventArgs>? LocationError;
```
I think .NET worlds use "failed"? So maybe 

```cs
event EventHandler<GeolocationListeningFailedEventArgs>? ListeningFailed;
```

If this works more with the listening things, then we should have listening. People might expect this to fire if any location things fail - like a normal GetLocationAsync.

mattleibow replied:
I see also the listen will throw an exception if something prevents the listener starting. So I am seeing that this event will only fire if something causes a problem _while_ a listening is happening - ie: permissions revoked.

I feel like we need "listening" in the name even more now.

vividos replied:
I renamed the event to `ListeningFailed`and the class to `GeolocationListeningFailedEventArgs`.

mattleibow commented on src/Essentials/src/Geolocation/Geolocation.shared.cs:
```diff
@@ -27,6 +29,40 @@ public interface IGeolocation
+		/// <summary>
+		/// Occurs while listening to location updates.
+		/// </summary>
+		event EventHandler<LocationEventArgs>? LocationChanged;
```
Not sure if this should be LocationChanged or LocationUpdated...

vividos replied:
It's up to you, either sounds good.

vividos replied:
@mattleibow If you have no preference I would let the event as is and just rename the event args class, as requested above.

mattleibow replied:
Sounds good.

mattleibow commented on src/Essentials/src/Geolocation/Geolocation.uwp.cs:
```diff
@@ -9,6 +9,13 @@ namespace Microsoft.Maui.Devices.Sensors
+		/// <summary>
+		/// Indicates if currently listening to location updates while the app is in foreground.
+		/// </summary>
+		public bool IsListeningForeground { get => false; }
```
Should this always be false here? I see code to do listening on this platform...

vividos replied:
You're right, I fixed it.

mattleibow reviewed (commented):
Very close! Just my usual nit picks with naming.

vividos committed:
renamed event LocationError to ListeningFailed and GeolocationErrorEventArgs to GeolocationListeningFailedEventArgs

vividos committed:
fixed IsListeningForeground property on Windows

jfversluis committed:
Merge branch 'main' into geolocation-foreground-listener

mattleibow reviewed (commented):
Looking very nice now. Just 2 things:

 - rename the event args class name
 - verify (either way) if this also works in unpackaged

jfversluis commented:
Just checked unpackaged which seems to work fine

jfversluis committed:
Fixed naming

jfversluis committed:
Update GeolocationViewModel.cs

Commit:
Thank you for your pull request. We are auto-formating your source code to follow our code guidelines.

jfversluis commented:
Thank you so much @vividos for bringing this to us! Really appreciate your time and effort with this one. Looking forward to your next contributions!

jfversluis commented:
@vividos btw, if you want, could you please fill out this survey about your experience? It will help us make it even better in the future! Thanks!

https://www.surveymonkey.com/r/8KTX9VT?Source=dotnet/maui

vividos commented:
Thanks Gerald and Matt for the review and the last commits!

alexeygritsenko commented:
Hi, as far as I understand the essence of this topic - you are migrating **jamesmontemagno/GeolocatorPlugin** to **.NET 8**. In this case, I would like to take into account this report https://stackoverflow.com/questions/69744671/problem-updating-location-every-5-seconds-with-xamarin-essentials Users write that the plugin is not accurate. Perhaps it contains some errors that need to be corrected.

vividos commented:
@alexeygritsenko thanks for the link. If you study the code, the foreground listening code is only a thin wrapper around Android's `RequestLocationUpdates()` which delivers new locations using the `ILocationListener` interface, which sends the location out using the `LocationChanged` event. Similar for the other platforms.
